createUser: mark emails as unverified on creation

bobheadxi opened the pull request:
Today, users created with `mutation { createUser }` automatically have their emails marked as verified, even if the email is bogus. This could cause us to automatically send potentially large numbers of emails automatically to nonexistent email addresses (see https://github.com/sourcegraph/customer/issues/1790). To mitigate this, this change marks the emails of created users as unverified if SMTP is configured.

This has implications for Cloud instances, where `mutation { createUser }` is frequently used to create initial admin users. After this change, these users may receive a more limited set of emails (i.e. https://github.com/sourcegraph/sourcegraph/pull/46184) and have limited capabilities (i.e. unable to link account via external service) until they verify their email by:

1. using a "set password" or "reset password" link delivered by email (https://github.com/sourcegraph/sourcegraph/pull/46307)
1. using "verify email" in user settings to send an email verification
3. a site admin verifying the user email manually via `setUserEmailVerified` mutation or UI

## Test plan

<!-- All pull requests REQUIRE a test plan: https://docs.sourcegraph.com/dev/background-information/testing_principles -->

I've added new unit tests that cover the old and new set password behaviour on the mutation.

Manual test:

1. `sg start`, configured with SMTP in `dev-private`
2. Site admin -> users -> create `testuser` with email 
3. Site admin -> users -> `testuser` -> emails -> primary email is unverified
4. Log out
5. CLick password reset link in email
6. Log in
7. User -> emails -> primary email is verified

bobheadxi commented:
Current dependencies on/for this PR:
* main
  * **PR #46183** <a href="https://app.graphite.dev/github/pr/sourcegraph/sourcegraph/46183" target="_blank"><img src="https://static.graphite.dev/graphite-32x32.png" alt="Graphite" width="10px" height="10px"/></a> 
    * **PR #46184** <a href="https://app.graphite.dev/github/pr/sourcegraph/sourcegraph/46184" target="_blank"><img src="https://static.graphite.dev/graphite-32x32.png" alt="Graphite" width="10px" height="10px"/></a> 
      * **PR #46187** <a href="https://app.graphite.dev/github/pr/sourcegraph/sourcegraph/46187" target="_blank"><img src="https://static.graphite.dev/graphite-32x32.png" alt="Graphite" width="10px" height="10px"/></a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/sourcegraph/sourcegraph/46187?utm_source=stack-comment).

bobheadxi commented on cmd/frontend/graphqlbackend/users_create.go:
```diff
@@ -29,34 +30,58 @@ func (r *schemaResolver) CreateUser(ctx context.Context, args *struct {
-		log15.Error("Failed to grant user pending permissions", "userID", user.ID, "error", err)
+		r.logger.Error("failed to grant user pending permissions",
+			log.Error(err))
 	}
 
```
we _could_ send the verification email here:

```go
	if verificationCode != "" && conf.CanSendEmail() {
		if err := backend.SendUserEmailVerificationEmail(ctx, user.Username, email, verificationCode); err != nil {
			return nil, errors.Wrap(err, "failed to send verification email")
		}
	}
```

bobheadxi replied:
Resolved in https://github.com/sourcegraph/sourcegraph/pull/46307

sg-e2e-regression-test-bob commented:
## Bundle size report 📦

| Initial size | Total size | Async size | Modules |
| --- | --- | --- | --- |
| 0.00% (+0.04 kb) | 0.00% (+0.63 kb) | 0.01% (+0.59 kb) | 0.00% (0) |

Look at the [Statoscope report](https://console.cloud.google.com/storage/browser/_details/sourcegraph_reports/statoscope-reports/01-05-email_do_not_send_emails_to_unverified_emails_split/compare-report.html;tab=live_object) for a full comparison between the commits 5b40d8f and 7a6dc29 or [learn more](https://docs.sourcegraph.com/dev/how-to/testing#bundlesize).


<details>
<summary>Open explanation</summary>

- `Initial size` is the size of the initial bundle (the one that is loaded when you open the page)
- `Total size` is the size of the initial bundle + all the async loaded chunks
- `Async size` is the size of all the async loaded chunks
- `Modules` is the number of modules in the initial bundle
</details>

unknwon commented:
> * We could [send a verification email on user creation](https://github.com/sourcegraph/sourcegraph/pull/46187#discussion_r1063020535), but these aren't needed if an admin expects a user to sign in via an external service, and would send an extra email on top of the password reset link if needed. This is also awkward because you must sign in before you click the verification.

IIRC, if an account is pre-created with an unverified and the user is expected to sign in via an external service, a brand new account is always created because we only consider account linking by _verified emails_.

unknwon commented on client/web/src/site-admin/SiteAdminCreateUserPage.tsx:
```diff
@@ -147,7 +165,7 @@ export class SiteAdminCreateUserPage extends React.Component<RouteComponentProps
                             />
                             <small id="site-admin-create-user-page__form-email-help" className="form-text text-muted">
-                                Optional verified email for the user.
+                                Optional email for the user - verification will be required if provided.
```
I find it rare to use `-` in in-product descriptions
```suggestion
                                Optional email for the user that will require verification when provided.
```

unknwon reviewed (commented):
Minor thinking: should we keep the current behavior if SMTP is not configured?

bobheadxi commented:
> Minor thinking: should we keep the current behavior if SMTP is not configured?

SMTP configuration is not static - you might not set up SMTP until some time after an instance is stood up, for example. Also, whether the email is verified or not doesn't impact instances that don't have SMTP set up, since we cannot send emails anyway

> IIRC, if an account is pre-created with an unverified and the user is expected to sign in via an external service, a brand new account is always created because we only consider account linking by verified emails. Not saying we can't change the current behavior, just want to call out the implication.

Ah, you are right. I think I'll look harder look at having the password reset also handle email verification before making this change then

kopancek commented on client/web/src/site-admin/SiteAdminCreateUserPage.tsx:
```diff
@@ -147,7 +165,7 @@ export class SiteAdminCreateUserPage extends React.Component<RouteComponentProps
                             />
                             <small id="site-admin-create-user-page__form-email-help" className="form-text text-muted">
-                                Optional verified email for the user.
+                                Optional email for the user that will require verification when provided.
```
This is a page for a site admin. I think they should be knowledgable enough to provide a non-bogus email for newly created users. So I think we should have this form create users with verified emails, as we did until now.

bobheadxi replied:
We recently had an admin create a user with a bogus email, so I'm not sure this is necessarily the case.

That said, we only _really_ care in Cloud - I could add this behind a feature flag instead

bobheadxi replied:
If a site admin wants to create a user and mark it as verified they can use the mutation `setUserEmailVerified`, which is also available in the UI

bobheadxi replied:
I've updated the text generated here such that the output creates a link to the user profile, for ease of access to the manual email verification. The behaviour has also been updated such that:

1. In the UI, if SMTP is enabled _and_ password resets are possible, we create with email unverified. This is because the user is very likely to click the password reset link via their email, which will verify the email.
2. In the API, the old behaviour is always retained.

kopancek commented on cmd/frontend/graphqlbackend/schema.graphql:
```diff
@@ -185,7 +185,7 @@ type Mutation {
         username: String!
         """
-        The new user's optional email address. If given, it is marked as verified.
+        The new user's optional email address. If given, it must be verified by the user.
```
Instead of saying this, I would add an optional boolean parameter to the graphql endpoint - `emailVerified` or similar. That way we can say on the graphql request if the email should be marked as verified or not. By default if the parameter is not present, the email would not be marked as verified.

Which should allow for the behavior that is needed - not marking emails as verified on our graphql requests.

Since this is a backwards incompatible change, I think it also needs to be mentioned on the changelog.

bobheadxi replied:
Added a changelog entry representing the change in default behaviour. IMO we shouldn't parameterize the behaviour, since there are ways a site admin can explicitly override it - continuing discussion in https://github.com/sourcegraph/sourcegraph/pull/46187#discussion_r1067861965

bobheadxi replied:
updated - see https://github.com/sourcegraph/sourcegraph/pull/46187#discussion_r1082096098

kopancek commented on cmd/frontend/graphqlbackend/users_create.go:
```diff
@@ -29,34 +30,59 @@ func (r *schemaResolver) CreateUser(ctx context.Context, args *struct {
 	}
 
-	// The new user will be created with a verified email address.
+	// 🚨 SECURITY: Do not assume user email is verified on creation if email delivery is
+	// enabled and we are allowed to reset passwords (which will become the primary
+	// mechanism for verifying this newly created email).
+	needsEmailVerification := email != "" && conf.CanSendEmail()
```
See my comment above. I think if we make the `emailVerified` param part of the request, the code here would need to be slightly modified.

kopancek commented on cmd/frontend/graphqlbackend/users_create.go:
```diff
@@ -66,22 +92,20 @@ func (r *createUserResult) ResetPasswordURL(ctx context.Context) (*string, error
 
-	var ru string
-	if conf.CanSendEmail() {
+	if conf.CanSendEmail() && r.needsEmailVerification {
```
I'm not sure this check is correct. 

Previous behavior: if we have SMTP configured, we send the reset password email to the user.

New behavior: if we have SMTP configured, we do not send the reset password email unless email needs to be verified. Seems like it would not work correctly for all the cases...

bobheadxi replied:
You're right, fixed!

kopancek reviewed (changes requested):
I think some of the changes do not make sense from a UX perspective. It still makes sense to let admins create a user with emails marked as verified - e.g. they know what they are doing and we should trust them.

I would like to also see some unit tests, since this is changing critical functionality (sending password reset links to newly created users).

bobheadxi commented on client/web/src/site-admin/SiteAdminCreateUserPage.tsx:
```diff
@@ -89,6 +89,31 @@ export class SiteAdminCreateUserPage extends React.Component<RouteComponentProps
     }
 
     public render(): JSX.Element | null {
+        const postCreateText = (result: CreateUserResult['createUser'], emailProvided: boolean): React.ReactNode => {
+            if (result.resetPasswordURL) {
+                const copyableURL = <CopyableText text={result.resetPasswordURL} size={40} />
+                if (window.context.emailEnabled && emailProvided) {
+                    return (
+                        <>
+                            <Text>
+                                A password reset URL has been sent to the new user's email address. If they don't
+                                receive it, you can also share the following password reset link:
+                            </Text>
+                            {copyableURL}
+                        </>
+                    )
+                }
+
+                return (
+                    <>
+                        <Text>You must manually send this password reset link to the new user:</Text>
+                        {copyableURL}
+                    </>
+                )
+            }
+            return <Text>The user must authenticate using a configured authentication provider.</Text>
+        }
```
nice :)

bobheadxi commented on cmd/frontend/graphqlbackend/users_create.go:
```diff
@@ -66,22 +92,20 @@ func (r *createUserResult) ResetPasswordURL(ctx context.Context) (*string, error
+		return &ru, nil
 	}
 
-	// This method modifies the DB, which is somewhat counterintuitive for a "value" type from an
-	// implementation POV. Its behavior is justified because it is convenient and intuitive from the
-	// POV of the API consumer.
```
Added it back 👍

bobheadxi commented on cmd/frontend/graphqlbackend/users_create.go:
```diff
@@ -29,34 +30,59 @@ func (r *schemaResolver) CreateUser(ctx context.Context, args *struct {
+	// mechanism for verifying this newly created email).
+	needsEmailVerification := email != "" && conf.CanSendEmail()
+	var emailVerificationCode string
+	if needsEmailVerification {
+		var err error
+		emailVerificationCode, err = backend.MakeEmailVerificationCode()
+		if err != nil {
+			return nil, errors.Wrap(err, "failed to generate email verification code")
+		}
+	}
```
A verification code is required to set `Verified: false`, so if it fails to generate then we can't create the user

bobheadxi commented on cmd/frontend/graphqlbackend/users_create.go:
```diff
@@ -29,34 +30,59 @@ func (r *schemaResolver) CreateUser(ctx context.Context, args *struct {
 		return nil, err
 	}
 
+	logger := r.logger.Scoped("createUser", "create user handler").With(
```
I've added more logging here

bobheadxi commented:
@kopancek thanks for the first round of reviews - I've addressed your comments and updated the behaviour, would appreciate a review when you get the chance!

mrnugget reviewed (commented):
Do we also need to update docs?

bobheadxi commented:
@mrnugget I've created some docs around email verification here: https://github.com/sourcegraph/sourcegraph/pull/46187/commits/00d5137881812fc280f7b7265df34aadcb2e28d3

mrnugget commented:
Thanks!

unknwon commented on cmd/frontend/internal/auth/userpasswd/set_password.go:
```diff
@@ -16,17 +16,7 @@ import (
-		return "", errors.Wrap(err, "get user by ID")
-	}
-
+func HandleSetPasswordEmail(ctx context.Context, db database.DB, id int32, username string, email string, emailVerified bool) (string, error) {
```
```suggestion
func HandleSetPasswordEmail(ctx context.Context, db database.DB, id int32, username, email string, emailVerified bool) (string, error) {
```

kopancek commented on CHANGELOG.md:
```diff
@@ -23,6 +23,7 @@ All notable changes to Sourcegraph are documented in this file.
 ### Changed
 
 - Code Insights has a new UI for the "Add or remove insights" view, which now allows you to search code insights by series label in addition to insight title. [#46538](https://github.com/sourcegraph/sourcegraph/pull/46538)
+- When SMTP is configured, users created by site admins via the "Create user" page will no longer have their email verified by default - users must verify their emails by using the "Set password" link they get sent, or have their emails verified by a site admin via the the "Emails" tab in user settings or the `setUserEmailVerified` mutation. The `createUser` mutation retains the old behaviour of  ...
```
Typo `the the "Emails" tab`
```suggestion
- When SMTP is configured, users created by site admins via the "Create user" page will no longer have their email verified by default - users must verify their emails by using the "Set password" link they get sent, or have their emails verified by a site admin via the "Emails" tab in user settings or the `setUserEmailVerified` mutation. The `createUser` mutation retains the old behaviour of automatically marking emails as verified. To learn more, refer to the [SMTP and email delivery](https://docs.sourcegraph.com/admin/config/email) documentation. [#46187](https://github.com/sourcegraph/sourcegraph/pull/46187)
```

kopancek commented on doc/admin/auth/index.md:
```diff
@@ -54,10 +54,20 @@ external identity provider.
 The [`builtin` auth provider](../config/site_config.md#builtin-password-authentication) manages user accounts internally in its own database. It supports user signup, login, and password reset (via email if configured, or else via a site admin).
 
-Password reset links expire after 4 hours.
+Password reset links expire after 4 hours by default - this can be configured in site configuration with the `auth.passwordResetLinkExpiry` field.
```
I think it would be better to have a link that points to the site config schema definition here. Not sure how feasible that is though 🤔

bobheadxi replied:
Added a link!

bobheadxi replied:
Trick seems to be to view the `site_config` page in browser, and get the anchor link through there

kopancek reviewed (approved):
Looks good to me. Added small sugestions that should not block PR merge.
